core-bundle/src/Cron/Cron.php

```diff
@@ -0,0 +1,176 @@
+
+class Cron
+{
+    /**
+     * @var string[]
+     */
+    public const INTERVALS = ['monthly', 'weekly', 'daily', 'hourly', 'minutely'];
+
+    /**
+     * @var Connection
+     */
+    protected $db;
+
+    /**
+     * @var LoggerInterface
+     */
+    protected $logger;
+
+    /**
+     * @var bool
+     */
+    protected $debug;
+
+    /**
+     * @var array
+     */
+    protected $crons = [];
+
+    public function __construct(Connection $db, bool $debug = false, LoggerInterface $logger = null)
+    {
+        $this->db = $db;
+        $this->debug = $debug;
+        $this->logger = $logger;
+    }
+
+    /**
+     * Add a cron service.
+     *
+     * @param object $service
+     */
+    public function addCronJob($service, string $method, string $interval, int $priority = 0, bool $cliOnly = false): void
+    {
+        if (!\in_array($interval, self::INTERVALS, true)) {
+            throw new \InvalidArgumentException(sprintf('Invalid interval "%s"', $interval));
+        }
+
+        $this->crons[$interval][$priority][] = [$service, $method, $cliOnly];
+    }
+
+    /**
+     * Run the registered Contao cron jobs.
+     *
+     * @param bool $cliOnly Whether the cli only crons should be run.
+     */
+    public function run(bool $cliOnly = false): void
+    {
+        // Do not run if the last execution was less than a minute ago
+        if ($this->hasToWait()) {
+            return;
+        }
+
+        $currentRuns = [];
+
+        // Store the current timestamps
+        $currentTimestamps = [
+            'monthly' => date('Ym'),
+            'weekly' => date('YW'),
+            'daily' => date('Ymd'),
+            'hourly' => date('YmdH'),
+            'minutely' => date('YmdHi'),
+        ];
+
+        // Get the timestamps from tl_cron
+        $lastRuns = $this->db->executeQuery("SELECT * FROM tl_cron WHERE name != 'lastrun'")->fetchAll();
+
+        foreach ($lastRuns as $lastRun) {
+            $currentRuns[$lastRun['name']] = $lastRun['value'];
+        }
+
+        // Create the database entries
+        foreach (self::INTERVALS as $interval) {
+            if (!isset($currentRuns[$interval])) {
+                $currentRuns[$interval] = 0;
+                $this->db->insert('tl_cron', ['name' => $interval, 'value' => 0]);
+            }
+        }
+
+        // Run the jobs
+        foreach (self::INTERVALS as $interval) {
+            $currentTimestamp = $currentTimestamps[$interval];
+
+            // Skip empty intervals and jobs that have been executed already
+            if (empty($this->crons[$interval]) || $currentRuns[$interval] === $currentTimestamp) {
+                continue;
+            }
+
+            // Update the database before the jobs are executed, in case one of them fails
+            $this->db->update('tl_cron', ['value' => $currentTimestamp], ['name' => $interval]);
+
+            // Add a log entry if in debug mode (see #4729)
+            if ($this->debug && null !== $this->logger) {
```

Comment:
Is there any reason why we only log in debug mode? That makes no sense to me. I know it looks like it was already present but we should re-discuss this. Imho we should always log. I can already filter out the irrelevant log entries by setting the minimum log level.

Reply:
Yeah I don't think I would log anything to the Contao table. In my opinion, logging should be as follows:

* A general log entry whenever the check runs on `debug` level so that you can debug whether calls are correct.
* A log entry only if at least one job successfully ran on `info` level.

Then you only get log entries on info level if actually something ran. But restricting it to `$this->debug` makes no sense to me.

Reply:
Fine with me. Only the cron jobs themselves should actually log into the Contao system log table if they want to, not the Cron service itself (which they are doing right now anyway, like _Purged the search cache_, etc.)

@leofeyer what do you think?

Reply:
Agreed.

Reply:
Done. I have also added another debug logging entry for when a `cli_only` cron job is skipped when the cron is not run via the `cron:run` command.